appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt

```diff
@@ -0,0 +1,277 @@
+package com.appcues.ui.presentation
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import androidx.lifecycle.viewModelScope
+import com.appcues.AppcuesCoroutineScope
+import com.appcues.LoggingLevel.NONE
+import com.appcues.action.ActionProcessor
+import com.appcues.action.ExperienceAction
+import com.appcues.analytics.ExperienceLifecycleEvent.StepInteraction.InteractionType
+import com.appcues.data.model.Experience
+import com.appcues.data.model.RenderContext
+import com.appcues.data.model.StepContainer
+import com.appcues.logging.Logcues
+import com.appcues.statemachine.State
+import com.appcues.statemachine.State.BeginningStep
+import com.appcues.statemachine.State.EndingStep
+import com.appcues.statemachine.StepReference
+import com.appcues.statemachine.StepReference.StepGroupPageIndex
+import com.appcues.ui.ExperienceRenderer
+import com.appcues.ui.presentation.AppcuesViewModel.UIState
+import com.appcues.ui.presentation.AppcuesViewModel.UIState.Dismissing
+import com.appcues.ui.presentation.AppcuesViewModel.UIState.Idle
+import com.appcues.ui.presentation.AppcuesViewModel.UIState.Rendering
+import com.google.common.truth.Truth.assertThat
+import io.mockk.called
+import io.mockk.coEvery
+import io.mockk.coVerify
+import io.mockk.every
+import io.mockk.excludeRecords
+import io.mockk.mockk
+import io.mockk.slot
+import io.mockk.verify
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.UnconfinedTestDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TestRule
+
+@OptIn(ExperimentalCoroutinesApi::class)
+internal class AppcuesViewModelTest {
+
+    @get:Rule
+    val rule: TestRule = InstantTaskExecutorRule()
+
+    private val renderContext: RenderContext = RenderContext.Modal
+
+    private val coroutineScope: AppcuesCoroutineScope = AppcuesCoroutineScope(Logcues(NONE))
+
+    private val experienceStates = MutableSharedFlow<State>(1)
+    private val experienceRenderer: ExperienceRenderer = mockk {
+        coEvery { getStateFlow(RenderContext.Modal) } returns experienceStates
+    }
+
+    private val actionProcessor: ActionProcessor = mockk(relaxed = true)
+
+    private val onDismiss: () -> Unit = mockk(relaxed = true)
+
+    private lateinit var viewModel: AppcuesViewModel
+
+    private val uiStates = arrayListOf<UIState>()
+
+    private lateinit var stateJob: Job
+
+    @Before
+    fun setup() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
```

Comment:
setting the main dispatcher to UnconfinedTestDispatcher allows us to use viewModelScope from ViewModel() (we do it on init block)